### DISCUSSION TITLE
specification: chunk index records for empty chunks are ambiguous

### DISCUSSION THREAD

**Discussion by wkalt** [foxglove/mcap#1573] [General]
The specification says that chunk index records for chunks with no messages should have timestamps set to zero. However, zero is a valid timestamp. Therefore it is impossible to distinguish an empty chunk from one that has one or more messages at time zero.

I think the nicest way to fix this would be to add a MessageCount field to the end of chunk indexes. This would resolve the ambiguity, and also leave the index records in place in case there is some useful thing other than messages that does go in the chunks in the future.

The new field would need to be optional until V1, at which point it could be made required.

**Comment by amacneil** [comment]
Do we have conformance tests or something to cover this across implementations?

**Comment by wkalt** [comment]
I backed out the implementation change and repurposed this ticket - but no, I think we probably do need to add a conformance test with an empty chunk.

**Comment by jtbandes** [comment]
Why is it important to distinguish empty chunks from chunks with messages at time zero?

**Comment by wkalt** [comment]
@jtbandes important is subjective but IMO,
* better to avoid ambiguous possibilities all things equal
* current situation means readers need to make an extra seek/range request for chunks that are actually empty
* current situation means there is no reliable way for a reader to determine the start time of a file, if the statistics record is not recorded but the chunk indexes are

**Comment by james-rms** [comment]
Changing this to a "feature" ticket, because it's hard to point to a tool or user experience that is degraded due to this issue.

Re: additional message count field - this seems reasonable to me. We're missing a lot of detail in the spec for how "optional" fields should work (or fields that were added after the initial release of the spec). Some questions that need answering are:

1. Are any fields in an MCAP record "required" ie a reader can consider it an error if they're missing? How do we label those?
2. Do we need well-defined default or zero values for all fields if they are not present?
